File: Source/NextTurnRuntime/Private/CoreUObject/DelegateMethod.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME DelegateMethod

TYPE_EXPORT_START

void EXPORT_CALL_CONV GetNameSuffix(const TCHAR*& SuffixPtr, int32_t& SuffixLength)
{
	SuffixPtr = HEADER_GENERATED_DELEGATE_SIGNATURE_SUFFIX;
	SuffixLength = UE_ARRAY_COUNT(HEADER_GENERATED_DELEGATE_SIGNATURE_SUFFIX) - 1;
}

const FMemberSymbol Members[] =
{
	EXPORT_METHOD(GetNameSuffix),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;

File: Managed/NextTurn.UE.Runtime/CoreUObject/DelegateMethod.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using System;
using NextTurn.UE.Annotations;

namespace Unreal
{
    public class DelegateMethod : Method
    {
        internal DelegateMethod(IntPtr pointer) : base(pointer) { }

        public static unsafe Span<char> NameSuffix
        {
            get
            {
                NativeMethods.GetNameSuffix(out void* suffixPtr, out int suffixLength);
                return new Span<char>(suffixPtr, suffixLength);
            }
        }

        private static new class NativeMethods
        {
            [Calli]
            public static extern unsafe void GetNameSuffix(out void* suffixPtr, out int suffixLength);
        }
    }
}
